core/silkworm/execution/evm.cpp

```diff
@@ -50,7 +50,8 @@ class DelegatingTracer : public evmone::Tracer {
         tracer_.on_instruction_start(pc, stack_top, stack_height, state, intra_block_state_);
     }
 
-    void on_execution_end(const evmc_result& result) noexcept override {
+    void on_execution_end(const evmc_result& res) noexcept override {
+        CallResult result{res.status_code, static_cast<uint64_t>(res.gas_left), {res.output_data, res.output_size}};
```

Comment:
Question ... why cast `res.gas_left` from int64_t to uint64_t ? I believe execution can return negative gas left which later raises an error of insufficient gas and casting might lead to martian value. Am I missing something ?

Reply:
Good catch. I saw the same conversion applied at the end of `EVM::execute` method, but I missed that both `EVM::create` and `EVM::call` take care to set the gas left to zero in case of any error except `EVMC_REVERT`.

So now the choice is between:
- applying the same reset of gas left to zero in case of any error except `EVMC_REVERT` also in `DelegateTracer::on_execution_end` before calling the tracer. This would allow to maintain the same `CallResult` parameter type for all `EvmTracer` methods carrying an execution result and also would keep `EvmTracer` consistent with the return type of `EVM::execute`
- reverting to use the previous result type, i.e. `evmc_result` in `EvmTracer` in order to propagate to the tracer implementations exactly what the `evmone` reported after execution

I slightly prefer the former in terms of interface coherence, but the latter seems more correct to allow any tracer implementation to log also negative gas left after execution, if needed.

What do you think?

Reply:
Sorry for not replying yesterday.
I agree data correctness is better than code coherence.

Reply:
OK done in last commit